### ISSUE TITLE
Upgrade to ESP32 - Wemos S2/C3 mini and S3

### ISSUE THREAD

**Issue by CheeseE** [emsesp/EMS-ESP32#605]
**Question**
Has anyone tried [this](https://www.wemos.cc/en/latest/s2/s2_mini.html) or [this](https://www.wemos.cc/en/latest/c3/c3_mini.html) board to upgrade the old ESP8266 to ESP32? Is it pin-to-pin compatible with the D1 mini? It seems to me but it would be good for someone to confirm, thanks.

**Comment by proddy** [comment]
The esp32-s2 is a different chip so the code needs to compiled. I've haven't tried it myself. 

https://docs.platformio.org/en/stable/boards/espressif32/lolin_s2_mini.html

**Comment by bbqkees** [comment]
It's the easiest to just get a MH-ET LIVE D1 Mini with ESP32 chip.
Pin compatible with the ESP8266 D1 Mini and it already has a board profile in the firmware.

**Comment by CheeseE** [comment]
Thank you, I have the MH-ET, just wanted to know if I have to cut the enclosure, as these mentioned boards are the same size as the D1 mini. Ordered them, and will try them out if it works.

**Comment by bbqkees** [comment]
The MH-ET is wider so you have to make a few adjustments to the housing. See the wiki:
https://bbqkees-electronics.nl/wiki/gateway/upgrade-the-gateway-to-an-esp32.html

**Comment by MichaelDvP** [comment]
If you dont want to cut the housing, you can also cut the board and sidepanel:
![PXL_20220826_121617078_1](https://user-images.githubusercontent.com/59284019/186883946-a644b298-ce6e-4ea1-8e10-ef0ad5933001.jpg)

![PXL_20220826_121556632_1](https://user-images.githubusercontent.com/59284019/186883956-f0e4c5c2-a14b-4c6b-9c60-f5cbfab812b6.jpg)

**Comment by CheeseE** [comment]
What I'm trying to avoid is to cut the housing. But this is a good idea to cut the board itself, thank you :)

**Comment by MichaelDvP** [comment]
You also have to remove the plastic spacer from the board pins to make the usb fit without cutting the housing.

**Comment by ajvdw** [comment]
I did build for S2: there are 3 minor adjustments necessary NO UART2, NO BT skip include esp_bt.h and I modified the rename_fw script to generate a S2 specific bin file. But the settings are not stored in NVS. It's already mentioned as an issue in 5.4.2. If that part is fixed I can do some testing. I will create an PR if it works.

**Comment by CheeseE** [comment]
@ajvdw  There's already a working version committed for Lolin C3 mini.

**Comment by ajvdw** [comment]
> @ajvdw There's already a working version committed for Lolin C3 mini.

I followed the same path ;-) 

I'd like to propose the bin files to follow the names in the PlatformIO Project environment, the rename_fw script would be something like, so no need to mention the flash size separately:

chip_target = env.get('PIOENV').upper() 

    print("app version: "+app_version)
    print("chip_target: "+chip_target)
    print("platform: "+platform)
    # print("flash size: "+flash_size)

    # convert . to _ so Windows doesn't complain
    variant = "EMS-ESP-" + app_version.replace(".", "_") + "-" + chip_target

**Comment by proddy** [comment]
That's a good recommendation. I had already change the file in the latest flash_optmization branch (which I'll merge into dev soon): https://github.com/emsesp/EMS-ESP32/blob/58a0ec9ccaefe5373044220507ce093096cde71b/scripts/rename_fw.py#L39 but your change is better. thanks

**Comment by MichaelDvP** [comment]
Today i've got a C3-Mini and a S2-Mini from lolin store. I'll check memory and issues the next days. S2 has less ram, but PSRAM. I'm curious how this works.

**Comment by proddy** [comment]
> PSRAM

I'm also curious how to use it (there are a few ways described [here](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/external-ram.html#configuring-external-ram)). I'm also going to start looking at how to measure performance using some kind of profile. There must be something out there that can monitor. The alternative is to build a test into test.cpp which will stress test EMS-ESP by loading data, printing it, publishing it, and repeating for a few minutes, and then logging CPU and memory consumption. There are 2 cores in the ESP32 standard and I don't think we're even using the 2nd one fully.

**Comment by MichaelDvP** [comment]
C3 is running now with ems, dallas, analog, had some issues with gpio settings. Program size is larger, free ram a bit less then esp32, but working good. 
S2 seems to have uart1 fixed on gpio17/18, not pin compatible with D1 mini and no plug in replacement for gateway premium.
psram is standard config as malloc, the dynamic json can use it and i think it is needed (without ems connection <50k free heap).
![grafik](https://user-images.githubusercontent.com/59284019/194710938-82d4e557-0cd1-4190-bc61-fc28ee96a13c.png)

S2 without ems, dallas, analog connection, program size for single core is smaller:
![grafik](https://user-images.githubusercontent.com/59284019/194711161-89938abd-3240-45f1-9e9e-7d9c32027137.png)

**Comment by MichaelDvP** [comment]
S2 uart pins have to be configured before driver loading. Now it works as plugin in gateway premium.
Psram is used without code change, no issues when calling api with large output. All working fast. 
Info with ems, dallas, analog connected:
![grafik](https://user-images.githubusercontent.com/59284019/194715004-97c5b55a-8de2-4fbc-89e4-6222e307c600.png)

**Comment by MichaelDvP** [comment]
@proddy To identify the platform in a string at central place i've modified the version.h to 
````
#define EMSESP_APP_VERSION "3.5.0b5"

if defined(ARDUINO_LOLIN_C3_MINI)
#define EMSESP_PLATFORM "ESP32-C3";
#else
#if defined(ARDUINO_LOLIN_S2_MINI)
#define EMSESP_PLATFORM "ESP32-S2";
#else
#define EMSESP_PLATFORM "ESP32";
#endif
#endif
```
Is this ok, or should i add a `platform.h` with this defines?

**Comment by proddy** [comment]
Yes! Nice.

**Comment by MichaelDvP** [comment]
It's better to check the chip, not the specific board. 
```
#if CONFIG_IDF_TARGET_ESP32C3
#define EMSESP_PLATFORM "ESP32-C3";
#elif CONFIG_IDF_TARGET_ESP32S2
#define EMSESP_PLATFORM "ESP32-S2";
#elif CONFIG_IDF_TARGET_ESP32
#define EMSESP_PLATFORM "ESP32";
#else
#error Target CONFIG_IDF_TARGET is not supported
#endif
```

The C3-Mini board have a rgb-LED (WS2812B), and can not be switchted by gpio, needs a fast protocoll with ns timing. Adafruit NeoLED lbrary can handle this, but flash is nearly full for ESP32C3. For now i skip LED.

**Comment by proddy** [comment]
I've ordered an S2 and C3 because I'm missing out on the fun too.

regarding the RGB, I have some old C code from another project years ago that handled the RGB LED on an ESP32, I'll see if I can reuse it.

**Comment by ajvdw** [comment]
I've done some testing with the S2, I can't change to GPIO 7 to 9 because of the validators. More GPIO's should be probably allowed. My board used GPIO18 and GP19 ont the ESP32 on the S2 the GPIO pins 7 and 9 are on that location.... I defined my board profile but I couldn't get it to connect to the EMS bus. I remember a remark about setting GPIO's before initializing the UART. But I can't find the matching commit with that fix.....

**Comment by proddy** [comment]
Those validators are easy to adjust for the different ESPs. I'm sure Michael can figure out the GPIOs init. I'm waiting for my S2's to arrive.

**Comment by MichaelDvP** [comment]
@ajvdw I will push a working dev soon to my repo and make a pr. Here is a working bin from my testings (with some extra debug logs, experimental values, etc.)
[EMS-ESP-LOLIN_S2_MINI-3_5_0b5.zip](https://github.com/emsesp/EMS-ESP32/files/9743455/EMS-ESP-LOLIN_S2_MINI-3_5_0b5.zip)

**Comment by ajvdw** [comment]
> @ajvdw I will push a working dev soon to my repo and make a pr. Here is a working bin from my testings (with some extra debug logs, experimental values, etc.) [EMS-ESP-LOLIN_S2_MINI-3_5_0b5.zip](https://github.com/emsesp/EMS-ESP32/files/9743455/EMS-ESP-LOLIN_S2_MINI-3_5_0b5.zip)

I've tested this version. It seems to work even with. the custom GPIO's set to 7 and 9.

**Comment by MichaelDvP** [comment]
The source is online now: [https://github.com/MichaelDvP/EMS-ESP32/tree/dev_](https://github.com/MichaelDvP/EMS-ESP32/tree/dev_).

**Comment by proddy** [comment]
Got my S2 Mini and C3 Mini today so will join the fun

**Comment by MichaelDvP** [comment]
There is a new MH-ET formfactor S3 board [LiLigo T7-S3](https://de.aliexpress.com/item/1005004777561826.html?spm=a2g0o.productlist.0.0.30f25aebjEipmY&algo_pvid=c5e83b49-5d41-447b-8ce7-41b9182cd816&algo_exp_id=c5e83b49-5d41-447b-8ce7-41b9182cd816-1&pdp_ext_f=%7B%22sku_id%22%3A%2212000030484765284%22%7D&pdp_npi=2%40dis%21EUR%2116.33%2113.06%21%21%21%21%21%402100bdd516674657192722403e0e7a%2112000030484765284%21sea&curPageLogUid=Af6pXVnaQQkC), 
A bit expensive actually, but 16MB flash, 8MB PSRAM, dualcore xtensa LX7 (liligo writes RISC-V, but S3 is LX7), that should be enough for some years.

**Comment by bbqkees** [comment]
The ESP32-S3-WROOM-1-N16R8 is pretty expensive, with double the component price of the ESP32 16MB.
But maybe a good one as a possible future successor.

**Comment by bbqkees** [comment]
I had a look at the datasheet. One benefit: it has USB built in. One bigger downside: It has ditched the convenient Ethernet MAC from the ESP32.

**Comment by MichaelDvP** [comment]
BTW: for testing something other i've connected the C3 buspowered parallel to my normal jackpowered ESP32.  On startup, while AP is active, C3 gives some incomplete rx, but after AP switches off, complet stable with no rx errors. This is a real recommendation for people wanting upgrade from v2 to v3 but needs buspower (no servicejack, no power-jack).

**Comment by bbqkees** [comment]
I'll update the ESP8266 upgrade page on the wiki to add the new chips.
Indeed if you are already on buspowered mode, being able to keep it is always nice.

**Comment by MichaelDvP** [comment]
Just a status message:
I have some issues with S2 and it's small internal RAM. Arduinojson uses the psram, ok, but it seems wifi/tcp needs some free internal ram and if mqtt queue grow (due to a misconfiguration) wifi gets disconnected. Same if i set weblog to 100 buffer. We should check the free heap for these queues. (I've also tested with different settings of CONFIG_SPIRAM_MALLOC_ALWAYSINTERNAL and CONFIG_SPIRAM_MALLOC_RESERVE_INTERNAL, but did not find how to change this at complile time)

**Comment by proddy** [comment]
ugh, that's a pity. I think doing a free heap check (using heap_caps_get_largest_free_block()) is the best way forward, but not on every addition as that would create a lot of overhead. Perhaps calculate the available heap every 10 seconds in the main loop. 

On my to-do list is also to see if using lighter-weight std libraries would ease up some memory. The std::string, map, array, queue eat up a lot of memory. In version 2 (ESP8266) I used https://github.com/muwerk/ustd for a while and it worked well.

**Comment by MichaelDvP** [comment]
I've tried with `ESP.getMaxAllocHeap() < 10240` on every publish and reading the heap only in system.healthcheck (5sec). Working, S2 is stable and wifi is reachable. But on my system it starts dropping mqtt messages on ~150 queue-size. No problem for me, but ha-autodiscovery  generates ~280 messages and ~130 of these are dropped. 
Better solution is to place all queues in psram, but don't know yet how to do this. BTW: web-log uses list and reducing the buffer size does not free mem, only indices are erased. Better to use deque.

**Comment by MichaelDvP** [comment]
Got the S3 https://www.lilygo.cc/products/t7-s3: Memory looks good (not connected to ems-bus). 
![Screenshot 2022-11-17 at 19-19-33 EMS-ESP](https://user-images.githubusercontent.com/59284019/202863008-b04fe8d6-9cb7-4f31-bda8-4829892a9998.png)

**Comment by proddy** [comment]
I've added S3 to the list - this is our next gen ESP32 chip to focus on

**Comment by MichaelDvP** [comment]
From discussions:

**Comment by tp1de** [comment]
Please re-open this issue for Liligo T7 S3.

With the Liligo S3 and v3.6.0-dev.7 I do have the following findings:

- Built in one of the older ems-esp boards from @bbqkees and connecting to service jack, the board runs instable. I believe voltage is too low - if provided by service jack and board. With usb3 power supply the WebUI and WiFi is stable.

- I do have the old board and the new one connected in parallel to the ems-bus with different bus-id's. The Liligo S3 is not capable of reading all telegrams - only approx. 10% are recognized compared to the other one. 

- Not all devices and only 20% of the entities are recognized. After some 10-20 minutes ems-esp is rebooting.

- EMS bus get's connected immediately. Quality is between 99 and 100% (due to the fact that 2 gateways are connected).   

- All functions are working. Just telegram processing, identifying entities and devices are not working properly. The whole computing process seems to run with very very low speed. 

- EMS-ESP Version v3.6.0-dev.7
Device (Platform / SDK)...........ESP32-S3 / v4.4.2
System Uptime............................000+00:02:39.115
CPU Frequency ..........................240 MHz...
Heap (Free / Max Alloc) .........188 KB / 179 KB
PSRAM (Size / Free) .................8.189 KB / 8.162 KB
Flash Chip (Size / Speed) ..... 16.384 KB / 80 MHz
Application (Used / Free) ......1.868 KB / 6.260 KB
File System (Used / Free) .......28 KB / 36 KB

**Comment by MichaelDvP** [comment]
Could you send a full log. 
Have you set different Bus-ids to the two ems-esp?  unplug the voltage regulator for usb-power?
I also see you are on SDK 4.4.2 (platform 5.2) , 
please add to your `pio_local.ini`
```
[env:lolin_s3]
platform = espressif32
build_flags = ${common.build_flags} -DEMSESP_DEBUG
```
then it's easier to compare to mine. Make the log after new compile to have the debug messages.
For compare, ET-MH32 on service jack, bus-id 0x0B, Liligo S3 on bus, usb powered, bus-id 0x12:
```
EMS-ESP Version v3.6.0-dev.7
Platform (Platform / SDK) ESP32-S3 / v4.4.4
System Betriebszeit 000+01:14:14.805
CPU Frequenz 240 MHz
freier RAM Speicher (Gesamt / max. Block) 192 KB / 167 KB
PSRAM (Größe / Frei) 8.189 KB / 8.169 KB
Flash Speicher (Größe / Geschwindigkeit) 16.384 KB / 80 MHz
Programm (Genutzt / Frei) 1.915 KB / 6.213 KB
Dateisystem (Genutzt / Frei) 28 KB / 36 KB
```
```
EMS-Telegramme empfangen (Rx) | 8.805 | 2 | 100%
EMS-Telegramme gelesen (Tx) | 1.162 | 2 | 100%
EMS-Telegramme geschrieben (Tx) | 0 | 0 |  
Temperatursensoren gelesen | 0 | 0 |  
Analogsensoren gelesen | 2.494 | 0 | 100%
MQTT-Nachrichten gesendet | 3.566 | 0 | 100%
API-Aufrufe | 0 | 0 |  
```

**Comment by tp1de** [comment]
@MichaelDvP The Lilygo S3 does not work properly. Compiling in debug modes makes the gateway not responding anymore.
Without debug flag I could access but response is incredible slow. E.g. looking at system log takes more then 20 seconds to display each line ....
I feel that I should flash with your flasher. Could you please explain how to do it.

**Comment by tp1de** [comment]
No different bus-id, s3 without mqtt, different hostname and ip-adresses. 
Sorry I can not display system log since it is basically not displayed and/or system reboots in the meantime. Even with just gateway (s3) it does not work. No chance to flash with your flasher?

**Comment by MichaelDvP** [comment]
I have tried the official build, it crashes every few seconds. Seems i have made a fix in my testbuild, have to check what is was.
Here is a bin of my testbuild. But it contains some changes and extras not tested completly( or not working at all).  But it works on my S3. 
[firmware.zip](https://github.com/emsesp/EMS-ESP32/files/11038272/firmware.zip)

**Comment by MichaelDvP** [comment]
Seems to be a stack overflow in UART task, change line 92 of `emsuart_esp32.cpp` to 
`xTaskCreate(uart_event_task, "uart_event_task", 2560, NULL, configMAX_PRIORITIES - 1, NULL);`
(change stack size from 2k to 2,5k) (my build uses 2,2k)

**Comment by tp1de** [comment]
@MichaelDvP yes uart stack size was causing the problems. 
Now S3 works, I just recognized one "misbehavior". When NTP switched on, I get every couple of seconds within system log the messages:
000+00:06:58.150 I 54: [thermostat] thermostat time correction from ntp
000+00:06:58.150 I 54: [thermostat] thermostat time correction from ntp
000+00:06:58.332 I 55: [thermostat] thermostat time correction from ntp
000+00:07:00.363 I 56: [thermostat] thermostat time correction from ntp
000+00:07:00.746 I 57: [thermostat] thermostat time correction from ntp
000+00:07:01.165 I 58: [thermostat] thermostat time correction from ntp
000+00:07:03.986 I 59: [thermostat] thermostat time correction from ntp

**Comment by tp1de** [comment]
Under S3 the 2 following thermostat entities are missing: hc1/hc2.haclimate. Anyhow I do have no glue what these entities are used for. Same code for S3 (16MB) and ESP32 (4MB) with new partioning.

**Comment by MichaelDvP** [comment]
This is a dummy entity to generate the homeassistant climate card depending on thermostat configuration. It should show up in settings customizations, but only has a value if HA is activated.

**Comment by MichaelDvP** [comment]
The esp32, s2, c3, s3 are different chip types and needs different binary files. The platformio.ini sets the right chiptype and partition size. 
For now you need to compile yourself, maybe later there will be different binaries for download.
The web upload checks file header and you can not upload a wrong binary and get http:503 service unavailable for wrong binaries.

Set in your  pio_local.ini 
```
[platformio]
 default_envs = esp32_16M, lolin_s3
```
to build both binaries if you hit build button.

**Comment by MichaelDvP** [comment]
Yes, S3 is working good, i have 2 gateways connected MH-ET and S3. C3 was also good, but not tested with the latest 3.6. With S2 i remember that memory was a bit critical, but maybe this was also the stack overflow from NTP service. My S2-tests are long time ago. We can reopen for S2 if someone use it an report issues.

**Comment by CheeseE** [comment]
Before we close, can we add the bin files to the release, so people can use them without compiling the whole stuff?

**Comment by bbqkees** [comment]
Well maybe now is a good time to start offering a bin file for S3 on the repository.

**Comment by bbqkees** [comment]
@CheeseE Seems we had the same idea at the same time :-)

**Comment by tp1de** [comment]
Same for me. Bin files for S3 would be fine. Is there any chance to provide old version bin files in the repository as well?

**Comment by MichaelDvP** [comment]
Tested the S2 again with actual v3.6dev. 
- Because of low heap mqtt is not working, we should check for PSRAM and heap to make it work again. But with HA enabled still ~20 mqtt errors happen while publishing the config. 
- I get some TX-read errors independend of tx-mode. After a tx-read it receives a poll and then the answer, If i ignore this poll when waiting for reply, all is well (zero tx-errors). The liligo S3 also shows this, but not as often as S2.

**Comment by proddy** [comment]
> Tested the S2 again with actual v3.6dev.
> 
> * Because of low heap mqtt is not working, we should check for PSRAM and heap to make it work again. But with HA enabled still ~20 mqtt errors happen while publishing the config.

Do you know what is causing the mqtt errors? Is it because of the heap check we added? I've been looking into https://github.com/bertmelis/espMqttClient which is new version of asyncMqttClient from one of the maintainers. It looks more lightweight and we can probably not use async and just send out the messages immediately because of the ESP32's dual core. I'm not sure though.

> * I get some TX-read errors independend of tx-mode. After a tx-read it receives a poll and then the answer, If i ignore this poll when waiting for reply, all is well (zero tx-errors). The liligo S3 also shows this, but not as often as S2.

Good catch. We can add that code to ignore it.

Also, do you think it's worth investing time into making EMS-ESP work on an S2?

**Comment by MichaelDvP** [comment]
> Do you know what is causing the mqtt errors?

Yes, mostly. The async-mqtt also checks for free heap (4k) and gives message id 0 back if lower. For the HA messages we can change the mqtt publish functions to bool, checking the message-Id and resend later if  not sent.
(Our heap check is 60k and has to be disabled for psram present, otherwise the s2 never sends)

>  I've been looking into https://github.com/bertmelis/espMqttClient which is new version of asyncMqttClient from one of the maintainers. It looks more lightweight and we can probably not use async and just send out the messages immediately because of the ESP32's dual core. I'm not sure though.

looks nice. Also the current async-mqtt has a queue and processes qos, works also direct sending. The ems-esp queue is only usefull to queue before connected. But the new lib looks good. We should test it.

> Also, do you think it's worth investing time into making EMS-ESP work on an S2?

No. But it's a nice test to see what happens if heap gets low and PSRam is active. Should be possible to make it stable.

**Comment by CheeseE** [comment]
@MichaelDvP could we add C3 binaries as well? I'm using that one for months now without issues.

**Comment by MichaelDvP** [comment]
The C3 works good, has more ram than S2 and with reduced tx-power buspowernig is possible. The Wemos C3 mini fits in the old gateway premium. It has only 4MB flash, but asymetric partitions with loader work.

But there are different versions of the wemos, v1.0.0 needs a special wifi init and have a single LED, v2.1.0 has a programable RGB-LED. So we need different binaries. I think it's better to compile yourself, or forc the repo and edit the github action to build your version. 

@proddy what do you think about adding C3 bnary?

**Comment by bbqkees** [comment]
I'll add an update to my wiki soon with instruction on selecting the correct bin file.

**Comment by MichaelDvP** [comment]
Just an update to modules:
Wemos/lolin have introduced a S3 mini https://www.wemos.cc/en/latest/s3/s3_mini.html with 4MB Flash and 2 MB PSRAM. But the lolin store is shut down https://lolin.aliexpress.com/ and the module can not be purcased https://de.aliexpress.com/item/1005005449219195.html Same for the C3 mini https://de.aliexpress.com/item/1005004740051202.html and S2, only S2 module is offered by different sellers on aliexpress. But S2 is to low with ram.
The liligo S3 module is available in liligo shop again https://www.lilygo.cc/products/t7-s3. Price is lower now, but no tax handling by lilgo for EU customers. On ebay the module is still 23-27€ from chinese resellers.

**Comment by proddy** [comment]
Thanks for sharing. Hopefully bbqkees has enough S3's in stock for his next batch of gateway boards. Then next I'd love to play with the C6 with its fast RISC-V and WiFi6 support.
